src/EventListener/DeserializeListener.php

```diff
@@ -39,28 +38,27 @@ final class DeserializeListener
 
     private $serializer;
     private $serializerContextBuilder;
-    private $formats = [];
+    private $formats;
     private $formatsProvider;
-    private $formatMatcher;
 
     /**
-     * @throws InvalidArgumentException
+     * @param ResourceMetadataFactoryInterface|FormatsProviderInterface|array $resourceMetadataFactory
      */
-    public function __construct(SerializerInterface $serializer, SerializerContextBuilderInterface $serializerContextBuilder, /* FormatsProviderInterface */$formatsProvider, ResourceMetadataFactoryInterface $resourceMetadataFactory = null)
+    public function __construct(SerializerInterface $serializer, SerializerContextBuilderInterface $serializerContextBuilder, $resourceMetadataFactory, ResourceMetadataFactoryInterface $legacyResourceMetadataFactory = null)
```

Comment:
We should remove the typehint for `$legacyResourceMetadataFactory`, right? And check that in the constructor body?